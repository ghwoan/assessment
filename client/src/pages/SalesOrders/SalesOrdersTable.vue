<template>
<div class="table-container">
  <div class="tcolumn thead">
        <div class="th">Order No</div>
        <div class="th" @click="sortBy('customer_name')" >Customer Name 
           <span class="arrow" :class="sortOrder.customer_name" ></span>
        </div>
        <div class="th" @click="sortBy('status')">Status 
           <span class="arrow" :class="sortOrder.status" ></span>
        </div>
        <div class="th" @click="sortBy('category')">Category 
           <span class="arrow" :class="sortOrder.category" ></span>
        </div>
        <div class="th" @click="sortBy('country')">Country 
           <span class="arrow" :class="sortOrder.country" ></span>
        </div>
        <div class="th" @click="sortBy('created_date')">Created Date 
           <span class="arrow" :class="sortOrder.created_date" ></span>
        </div>
  </div>

    <div v-if="orders && orders.length">    
      <div v-for="(order,index) in sortedData" :key="index" class="tcolumn trow"  :class="(index%2==0)? 'even' : 'odd'">
        <div  class="tcell left">{{index +1}}</div>
        <div class="tcell left">{{order.customer_name}}</div>
        <div class="tcell left">{{order.status}}</div>
        <div class="tcell left">{{order.category}}</div>
        <div class="tcell left">{{order.country}}</div>
        <div class="tcell left">{{formatDate(order.created_date)}}</div>
      </div>
    </div>
    <div v-else>
        <div class="tcell">
          No orders
        </div>
    </div>
  </div>

</template>

<script>
export default {
  props: {
    data: []
  },
  data() {
      return {
         orders: [],
         sortField: '',
         sortOrder: {customer_name:"asc", status:"asc", category: "asc", country:"asc", created_date: "asc"} 
      };
   },
  watch: {
      data: function (value) {
       //  console.log(value);
         this.orders = value;
      },
  },
  computed: {
      sortedData: function () {
        if (!this.sortField) return this.orders;
        return this.sortOrdersByString(this.sortField, this.sortOrder[this.sortField]);
     },
  },
   methods: {
      sortBy: function(fieldName){
        this.sortField = fieldName;
        if(this.sortOrder[fieldName]=="asc"){
          this.sortOrder[fieldName] = "dsc";
        }else{
          this.sortOrder[fieldName] = "asc";
        }
        },
    // to format the date to display format
      formatDate: function(date) {
        if(!date){
          return '';
        }
        try{
          const d = new Date(date);
          let strDate = d.toLocaleDateString();
          return strDate;
        }catch(er){
          return '';
        }
      }
      ,
      sortOrdersByString(fieldName, sortOrder) {
      
         let sorted = [...this.orders].sort((a, b) => {
            let al = a[fieldName].toLowerCase();
            let bl = b[fieldName].toLowerCase();
            
            if (al < bl) {
               return -1;
            }
            if (al > bl) {
               return 1;
            }
            return 0;
         });
         if (sortOrder!="asc"){
          sorted = sorted.reverse();
         }
         return sorted;
      }


   }
}
</script>

<style scoped>

.table-container {
  display: grid;
  grid-template-rows: auto;

  padding: 10px;
}
.tcolumn {
  display: grid;
  grid-template-columns: 10% 20% 15% 20% 15% 20%;
}
.thead {
  background-color:rgb(12, 41, 60);
  color:rgb(255,255,255);
  font-size:11pt
}
.th{
    padding: 8px;
  border: 1px solid rgba(180, 180, 180, 0.8);
}
.trow {
  padding: 0;
  color:rgb(25,25,25);
  font-size:11pt
}

.trow.even{
    background-color:rgb(242, 245, 250);
}
.trow.odd{
    background-color:rgb(255, 255, 255);
}
.tcell {
   padding: 5px;
}
.tcell.left{
  text-align: left;
  padding-left:10px;
  border: 1px solid rgba(180, 180, 180, 0.8);

}
.arrow {
  display: inline-block;
  vertical-align: middle;
  width: 0;
  height: 0;
  margin-left: 5px;
  opacity: 0.66;
}

.arrow.asc {
  border-left: 4px solid transparent;
  border-right: 4px solid transparent;
  border-bottom: 4px solid #fff;
}

.arrow.dsc {
  border-left: 4px solid transparent;
  border-right: 4px solid transparent;
  border-top: 4px solid #fff;
}
</style>

