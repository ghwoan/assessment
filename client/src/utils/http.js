import axios from 'axios';

const defConfig = {
  baseURL: process.env.VUE_APP_SERVER_URL || `http://localhost:8080/`,
  headers: {
    Authorization: process.env.VUE_APP_API_KEY || 'Testing'
  }
};

export const http = axios.create(defConfig);