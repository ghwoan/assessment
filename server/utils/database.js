import mysql from 'mysql2';
import dbConfig from "../config/db.config.js";

const pool = mysql.createPool({
    host: dbConfig.host,
    user: dbConfig.USER,
    database: dbConfig.DB,
   password: dbConfig.PASSWORD,
   waitForConnections: true,
   connectionLimit: 20,
});

const db = pool.promise();

export default db;