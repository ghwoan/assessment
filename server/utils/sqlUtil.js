
export const createDateRangeFilter = (fieldName, prefix, startDate, endDate) => {
   let sql = "";
   if (startDate) {
      sql = fieldName + " >= '" + startDate + "'";
   }
   if (endDate) {
      if (sql) {
         sql = sql + " AND ";
      }
      sql = sql + fieldName + " <='" + endDate + "'";
   }
   if (sql) {
      sql = " (" + sql + ")";
      if (prefix) {
         sql = prefix + " AND " + sql;
      }
   } else {
      sql = prefix;
   }
   return sql;
}

//define enum for FieldType
export const FieldTypeEnum = {
   NUMBER : 1,
   STRING : 2
}

export const escapeChar = (value) => {
   if (!value) {
      return value;
   } else {
      return "'" + value.split("'").join("\\'") + "'";
   }
}

export const createOpFilter = (fieldName, prefix, opFilter, fieldType) => {
   let sql = prefix;
   let items = "";
  
   if (!opFilter || !opFilter.list) {
      return prefix;
   };
   console.log(opFilter.list);
   if (opFilter.list.length) {
      opFilter.list.forEach(c => {
         if (items) items = items + ",";
         if (fieldType == FieldTypeEnum.NUMBER) {
            items = items + c;   
         } else {
            items = items +  escapeChar(c) ;
         }
      });
   }
   if (items) {
      items = "(" + items + ")";
      if (opFilter.op == "IN") {
         items = fieldName + " IN " + items;
      } else[
         items = fieldName + " NOT IN " + items
      ]
      sql = sql  + " AND (" + items + ")";
   }
   return sql;
}