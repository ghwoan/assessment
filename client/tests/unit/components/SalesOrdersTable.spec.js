import { http } from '@/utils/http';
import { shallowMount, mount } from '@vue/test-utils';
import SalesOrdersTable from '../../../src/components/SalesOrdersTable/SalesOrdersTable.vue';


describe('SalesOrdersTable.vue', () => {
  it('renders empty', () => {
    const msg = 'Order No Customer Name Status Category Country Created Date  No orders';
    const wrapper = shallowMount(SalesOrdersTable, {
      props: {  data: undefined}
    })
    //console.log(wrapper.html());
    // expect(wrapper.text()).toMatch(msg);
     expect(wrapper.html()).toMatchSnapshot();
  })

  it('renders data', async () => {
   const msg = 'Order No Customer Name Status Category Country Created Date  No orders';
   const wrapper = mount(SalesOrdersTable, {
     props: {  data: undefined}
   })
   await wrapper.setData({orders: [{orderId:"1",customerName: "B", country: "Indonesia" }, {orderId:"2", customerName:"A", country:"China" }]});
   expect(wrapper.html()).toMatchSnapshot();
 })

  it('method formatDate', async () => {
    const wrapper = await mount(SalesOrdersTable, {});
   // expect(wrapper.text()).toMatch(msg);
     let strDate = "2020-05-01";
     expect(wrapper.vm.formatDate(strDate)).toBe("01/05/2020");
     strDate = "2020-05-01T00:21:21";
     expect(wrapper.vm.formatDate(strDate)).toBe("01/05/2020");
     strDate = "2020-05-01 00:21:21";
     expect(wrapper.vm.formatDate(strDate)).toBe("01/05/2020");
     strDate = "2020/05/01";
     expect(wrapper.vm.formatDate(strDate)).toBe("01/05/2020");
     strDate = "";
     expect(wrapper.vm.formatDate(strDate)).toBe("");
  })
   
   
   it('method sortOrdersByString', async () => {
      const wrapper = await mount(SalesOrdersTable, {});
      await wrapper.setData({orders: [{orderId:"1", customerName: "Indonesia" }, {orderId:"2", customerName: "China" }]});
  
      let output = wrapper.vm.sortOrdersByString("customerName", "asc");
      expect(output.reduce((t, item) => { return t + item.customerName }, "")).toBe("ChinaIndonesia");
      expect(JSON.stringify(output)).toMatchSnapshot();
 
      output = wrapper.vm.sortOrdersByString("customerName", "dsc");
      expect(output.reduce((t, item) => { return t + item.customerName }, "")).toBe("IndonesiaChina");
      expect(JSON.stringify(output)).toMatchSnapshot();
 
      output = wrapper.vm.sortOrdersByString("customerName", "Abc");
      expect(output.reduce((t, item) => { return t + item.customerName }, "")).toBe("IndonesiaChina");
      expect(JSON.stringify(output)).toMatchSnapshot();
       
      output = wrapper.vm.sortOrdersByString("customerName", undefined);
      expect(output.reduce((t, item) => { return t + item.customerName }, "")).toBe("IndonesiaChina");
      expect(JSON.stringify(output)).toMatchSnapshot();

      output = wrapper.vm.sortOrdersByString("customerName", "");
      expect(output.reduce((t, item) => { return t + item.customerName }, "")).toBe("IndonesiaChina");
      expect(JSON.stringify(output)).toMatchSnapshot();
 
      output = wrapper.vm.sortOrdersByString("orderId", "");
      expect(output.reduce((t, item) => { return t + item.orderId }, "")).toBe("21");
      expect(JSON.stringify(output)).toMatchSnapshot();

      await wrapper.setData({ orders: [{ char: "a"}, { char: "D"}, { char: "c"}, { char: "b"}, { char: "A"}] });
      output = wrapper.vm.sortOrdersByString("char", "asc");
      expect(output.reduce((t, item) => { return t + item.char }, "")).toBe("aAbcD");

      output = wrapper.vm.sortOrdersByString("char");
      expect(output.reduce((t, item) => { return t + item.char }, "")).toBe("DcbAa");
   })

})
