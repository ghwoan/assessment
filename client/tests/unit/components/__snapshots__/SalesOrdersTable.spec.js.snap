// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SalesOrdersTable.vue method sortOrdersByString 1`] = `"[{\\"orderId\\":\\"2\\",\\"customerName\\":\\"China\\"},{\\"orderId\\":\\"1\\",\\"customerName\\":\\"Indonesia\\"}]"`;

exports[`SalesOrdersTable.vue method sortOrdersByString 2`] = `"[{\\"orderId\\":\\"1\\",\\"customerName\\":\\"Indonesia\\"},{\\"orderId\\":\\"2\\",\\"customerName\\":\\"China\\"}]"`;

exports[`SalesOrdersTable.vue method sortOrdersByString 3`] = `"[{\\"orderId\\":\\"1\\",\\"customerName\\":\\"Indonesia\\"},{\\"orderId\\":\\"2\\",\\"customerName\\":\\"China\\"}]"`;

exports[`SalesOrdersTable.vue method sortOrdersByString 4`] = `"[{\\"orderId\\":\\"1\\",\\"customerName\\":\\"Indonesia\\"},{\\"orderId\\":\\"2\\",\\"customerName\\":\\"China\\"}]"`;

exports[`SalesOrdersTable.vue method sortOrdersByString 5`] = `"[{\\"orderId\\":\\"1\\",\\"customerName\\":\\"Indonesia\\"},{\\"orderId\\":\\"2\\",\\"customerName\\":\\"China\\"}]"`;

exports[`SalesOrdersTable.vue method sortOrdersByString 6`] = `"[{\\"orderId\\":\\"2\\",\\"customerName\\":\\"China\\"},{\\"orderId\\":\\"1\\",\\"customerName\\":\\"Indonesia\\"}]"`;

exports[`SalesOrdersTable.vue renders data 1`] = `
<div class="table-container">
  <div class="tcolumn thead">
    <div class="th">Order No <span class="arrow asc"></span></div>
    <div class="th">Customer Name <span class="arrow asc"></span></div>
    <div class="th">Status <span class="arrow asc"></span></div>
    <div class="th">Category <span class="arrow asc"></span></div>
    <div class="th">Country <span class="arrow asc"></span></div>
    <div class="th">Created Date <span class="arrow asc"></span></div>
  </div>
  <div>
    <div class="tcolumn trow even">
      <!--div  class="tcell left">{{index +1}}</div-->
      <div class="tcell left">1</div>
      <div class="tcell left">B</div>
      <div class="tcell left"></div>
      <div class="tcell left"></div>
      <div class="tcell left">Indonesia</div>
      <div class="tcell left"></div>
    </div>
    <div class="tcolumn trow odd">
      <!--div  class="tcell left">{{index +1}}</div-->
      <div class="tcell left">2</div>
      <div class="tcell left">A</div>
      <div class="tcell left"></div>
      <div class="tcell left"></div>
      <div class="tcell left">China</div>
      <div class="tcell left"></div>
    </div>
  </div>
</div>
`;

exports[`SalesOrdersTable.vue renders empty 1`] = `
<div class="table-container">
  <div class="tcolumn thead">
    <div class="th">Order No <span class="arrow asc"></span></div>
    <div class="th">Customer Name <span class="arrow asc"></span></div>
    <div class="th">Status <span class="arrow asc"></span></div>
    <div class="th">Category <span class="arrow asc"></span></div>
    <div class="th">Country <span class="arrow asc"></span></div>
    <div class="th">Created Date <span class="arrow asc"></span></div>
  </div>
  <div>
    <div class="tcell"> No orders </div>
  </div>
</div>
`;
